docs/api/custom-integration.md

```diff
@@ -0,0 +1,115 @@
+# Custom integration
+
+!!! warning
+    This page is based on Ktlint `0.49.x` which has to be released. Most concepts are also applicable for `0.48.x`.  
+
+## Ktlint Rule Engine
+
+The `Ktlint Rule Engine` is the central entry point for custom integrations with the `Ktlint API`. See [basic API Consumer](https://github.com/pinterest/ktlint/blob/master/ktlint-api-consumer/src/main/kotlin/com/example/ktlint/api/consumer/KtlintApiConsumer.kt) for a basic example on how to invoke the `Ktlint Rule Engine`. This example also explains how the logging of the `Ktlint Rule Engine` can be configured to your needs.
+
+The `KtLintRuleEngine` instance only needs to be created once for the entire lifetime of your application. Reusing the same instance results in better performance due to caching. 
+
+```kotlin title="Creating the KtLintRuleEngine"
+val ktLintRuleEngine =
+    KtLintRuleEngine(
+        ruleProviders = KTLINT_API_CONSUMER_RULE_PROVIDERS,
+    )
+```
+
+### Rule provider
+
+The `KtLintRuleEngine` must be configured with at least one `RuleProvider`. A `RuleProvider` is a lambda which upon request of the `KtLintRuleEngine` provides a new instance of a specific rule. You can either provide any of the standard rules provided by KtLint or with your own custom rules, or with a combination of both.
+```kotlin title="Creating a set of RuleProviders"
+val KTLINT_API_CONSUMER_RULE_PROVIDERS =
```

Comment:
To get just the RuleProviders from Ktlint, it is more readable to use `StandardRuleSetProvider().getRuleProviders()` (see https://github.com/pinterest/ktlint/blob/0.48.2/ktlint-ruleset-standard/src/main/kotlin/com/pinterest/ktlint/ruleset/standard/StandardRuleSetProvider.kt). 

The code block above retrieves all rules which are defined in the project or in any of the dependencies that are included in the project. Standard rules will then only be retrieved in case the Ktlint dependency is included in which the Standard rules are defined.

In '0.48.x', experimental rules can be retrieved in a similar way as the standard rules. Note that in `0.49.x` the experimental rules will be included in the standard rules. See changelog for more information.